File: python-sdk/expansion/interfaces/rpc/shipyard.py
from typing import Optional, NamedTuple, Callable
from enum import Enum

import expansion.protocol.Protocol_pb2 as public
import expansion.protocol as protocol
from expansion.transport import IOTerminal, Channel
from expansion import utils


class Specification(NamedTuple):
    labor_per_sec: float


class ShipyardI(IOTerminal):
    class Status(Enum):
        SUCCESS = "success"
        INTERNAL_ERROR = "internal error"
        CARGO_NOT_FOUND = "cargo not found"
        BUILD_STARTED = "build started"
        BUILD_IN_PROGRESS = "build in progress"
        BUILD_COMPLETE = "build complete"
        BUILD_CANCELED = "build canceled"
        BUILD_FROZEN = "build frozen"
        BUILD_FAILED = "build failed"
        BLUEPRINT_NOT_FOUND = "blueprint not found"
        SHIPYARD_IS_BUSY = "shipyard is busy"
        # Internal SDK statuses:
        FAILED_TO_SEND_REQUEST = "failed to send request"
        RESPONSE_TIMEOUT = "response timeout"
        UNEXPECTED_RESPONSE = "unexpected response"
        CHANNEL_CLOSED = "channel closed"
        CANCELED = "operation canceled"

        @staticmethod
        def from_protobuf(status: public.IResourceContainer.Status) -> "ShipyardI.Status":
            ProtobufStatus = public.IShipyard.Status
            ModuleStatus = ShipyardI.Status
            return {
                ProtobufStatus.SUCCESS: ModuleStatus.SUCCESS,
                ProtobufStatus.INTERNAL_ERROR: ModuleStatus.INTERNAL_ERROR,
                ProtobufStatus.CARGO_NOT_FOUND: ModuleStatus.CARGO_NOT_FOUND,
                ProtobufStatus.BUILD_STARTED: ModuleStatus.BUILD_STARTED,
                ProtobufStatus.BUILD_IN_PROGRESS: ModuleStatus.BUILD_IN_PROGRESS,
                ProtobufStatus.BUILD_COMPLETE: ModuleStatus.BUILD_COMPLETE,
                ProtobufStatus.BUILD_CANCELED: ModuleStatus.BUILD_CANCELED,
                ProtobufStatus.BUILD_FROZEN: ModuleStatus.BUILD_FROZEN,
                ProtobufStatus.BUILD_FAILED: ModuleStatus.BUILD_FAILED,
                ProtobufStatus.BLUEPRINT_NOT_FOUND: ModuleStatus.BLUEPRINT_NOT_FOUND,
                ProtobufStatus.SHIPYARD_IS_BUSY: ModuleStatus.SHIPYARD_IS_BUSY,
                # To be extended
            }[status]

    BuildingReportCallback = Callable[[Status, Optional[float]], None]

    def __init__(self, name: Optional[str] = None):
        if not name:
            name = utils.generate_name(ShipyardI)
        super().__init__(name=name)

    @Channel.return_on_close(Status.CHANNEL_CLOSED, None)
    async def get_specification(self, timeout: float = 0.5) \
            -> (Status, Optional[Specification]):
        request = public.Message()
        request.shipyard.specification_req = True
        if not self.send(message=request):
            return ShipyardI.Status.FAILED_TO_SEND_REQUEST, None
        response, _ = await self.wait_message(timeout=timeout)
        if not response:
            return ShipyardI.Status.RESPONSE_TIMEOUT, None
        spec = protocol.get_message_field(response, ["shipyard", "specification"])
        if not spec:
            return ShipyardI.Status.UNEXPECTED_RESPONSE, None
        spec = Specification(labor_per_sec=spec.labor_per_sec)
        return ShipyardI.Status.SUCCESS, spec

    @Channel.return_on_close(Status.CHANNEL_CLOSED)
    async def bind_to_cargo(self, cargo_name: str, timeout: float = 0.5) -> Status:
        request = public.Message()
        request.shipyard.bind_to_cargo = cargo_name
        if not self.send(message=request):
            return ShipyardI.Status.FAILED_TO_SEND_REQUEST
        response, _ = await self.wait_message(timeout=timeout)
        if not response:
            return ShipyardI.Status.RESPONSE_TIMEOUT
        status = protocol.get_message_field(response, ["shipyard", "bind_to_cargo_status"])
        if status is None:
            return ShipyardI.Status.UNEXPECTED_RESPONSE
        return ShipyardI.Status.from_protobuf(status)

    @Channel.return_on_close(Status.CHANNEL_CLOSED)
    async def start_build(self, blueprint: str, ship_name: str) -> Status:
        request = public.Message()
        req_body = request.shipyard.start_build
        req_body.blueprint_name = blueprint
        req_body.ship_name = ship_name
        if not self.send(message=request):
            return ShipyardI.Status.FAILED_TO_SEND_REQUEST
        status, _ = await self.wait_building_report()
        return status

    @Channel.return_on_close(Status.CHANNEL_CLOSED, None)
    async def wait_building_report(self, timeout: float = 1.0) -> \
            (Status, Optional[float]):
        response, _ = await self.wait_message(timeout=timeout)
        if not response:
            return ShipyardI.Status.RESPONSE_TIMEOUT, None
        report = protocol.get_message_field(response, ["shipyard", "building_report"])
        if not report:
            return ShipyardI.Status.UNEXPECTED_RESPONSE, None
        return ShipyardI.Status.from_protobuf(report.status), report.progress

    @Channel.return_on_close(Status.CHANNEL_CLOSED, None, None)
    async def wait_building_complete(self, timeout: float = 1.0) -> \
            (Status, Optional[str], Optional[int]):
        response, _ = await self.wait_message(timeout=timeout)
        if not response:
            return ShipyardI.Status.RESPONSE_TIMEOUT, None, None
        report = protocol.get_message_field(response, ["shipyard", "building_complete"])
        if not report:
            return ShipyardI.Status.UNEXPECTED_RESPONSE, None, None
        return ShipyardI.Status.SUCCESS, \
               report.ship_name, \
               report.slot_id

File: typescript-sdk/lowlevel/shipyard.ts
import { create } from "@bufbuild/protobuf";
import * as msg from "../Protocol_pb.js";
import * as types from "../types/index.js";
import { Session } from "./session.js";


export type ShipyardStatus =
    | "SUCCESS"
    | "INTERNAL_ERROR"
    | "CARGO_NOT_FOUND"
    | "SHIPYARD_IS_BUSY"
    | "BUILD_STARTED"
    | "BUILD_IN_PROGRESS"
    | "BUILD_COMPLETE"
    | "BUILD_CANCELED"
    | "BUILD_FROZEN"
    | "BUILD_FAILED"
    | "BLUEPRINT_NOT_FOUND";

export type ShipyardSpecification = {
    labor_per_sec: number;
}

export type ShipyardBuildingReport = {
    status: ShipyardStatus;
    progress: number;
}

export type ShipyardShipBuilt = {
    ship_name: string;
    slot_id: number;
}

export type ShipyardBuildingEvent =
    | { case: "building_report"; report: ShipyardBuildingReport }
    | { case: "building_complete"; ship: ShipyardShipBuilt };

export class Shipyard {

    constructor(private session: Session) {}

    async send_specification_request(): Promise<types.Status> {
        const request = create(msg.IShipyardSchema, {
            choice: { case: "specificationReq", value: true },
        });
        return this.send(request);
    }

    async wait_specification(timeout: number = 500)
    : Promise<[types.Status, ShipyardSpecification | undefined]>
    {
        const [status, response] = await this.wait(timeout);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case != "specification") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        return [types.Status.ok(), {
            labor_per_sec: response.choice.value.laborPerSec,
        }];
    }

    async send_bind_to_cargo(cargo_name: string): Promise<types.Status> {
        const request = create(msg.IShipyardSchema, {
            choice: { case: "bindToCargo", value: cargo_name },
        });
        return this.send(request);
    }

    async wait_bind_to_cargo_status(timeout: number = 500)
    : Promise<[types.Status, ShipyardStatus | undefined]>
    {
        const [status, response] = await this.wait(timeout);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case != "bindToCargoStatus") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        const server_status = shipyardStatusFromProtobuf(response.choice.value);
        if (!server_status) {
            return [types.Status.fail(`got unexpected status ${response.choice.value}`),
                    undefined];
        }
        return [types.Status.ok(), server_status];
    }

    async send_start_build(blueprint_name: string, ship_name: string): Promise<types.Status> {
        const request = create(msg.IShipyardSchema, {
            choice: {
                case: "startBuild",
                value: {
                    blueprintName: blueprint_name,
                    shipName: ship_name,
                },
            },
        });
        return this.send(request);
    }

    async send_cancel_build(): Promise<types.Status> {
        const request = create(msg.IShipyardSchema, {
            choice: { case: "cancelBuild", value: true },
        });
        return this.send(request);
    }

    async wait_building_event(timeout: number = 1000)
    : Promise<[types.Status, ShipyardBuildingEvent | undefined]>
    {
        const [status, response] = await this.wait(timeout);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case == "buildingComplete") {
            const ship = response.choice.value;
            return [types.Status.ok(), {
                case: "building_complete",
                ship: {
                    ship_name: ship.shipName,
                    slot_id: ship.slotId,
                },
            }];
        }
        if (response.choice.case != "buildingReport") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        const report = response.choice.value;
        const server_status = shipyardStatusFromProtobuf(report.status);
        if (!server_status) {
            return [types.Status.fail(`got unexpected status ${report.status}`),
                    undefined];
        }
        return [types.Status.ok(), {
            case: "building_report",
            report: {
                status: server_status,
                progress: report.progress,
            },
        }];
    }

    private async send(request: msg.IShipyard): Promise<types.Status> {
        const message = create(msg.MessageSchema, {
            choice: { case: "shipyard", value: request },
        });
        return this.session.send(message);
    }

    private async wait(timeout_ms: number = 500)
    : Promise<[types.Status, msg.IShipyard | undefined]>
    {
        const [status, response] = await this.session.wait(timeout_ms);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case != "shipyard") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        return [types.Status.ok(), response.choice.value];
    }

}

function shipyardStatusFromProtobuf(
    value: msg.IShipyard_Status): ShipyardStatus | undefined
{
    switch (value) {
        case msg.IShipyard_Status.SUCCESS: return "SUCCESS";
        case msg.IShipyard_Status.INTERNAL_ERROR: return "INTERNAL_ERROR";
        case msg.IShipyard_Status.CARGO_NOT_FOUND: return "CARGO_NOT_FOUND";
        case msg.IShipyard_Status.SHIPYARD_IS_BUSY: return "SHIPYARD_IS_BUSY";
        case msg.IShipyard_Status.BUILD_STARTED: return "BUILD_STARTED";
        case msg.IShipyard_Status.BUILD_IN_PROGRESS: return "BUILD_IN_PROGRESS";
        case msg.IShipyard_Status.BUILD_COMPLETE: return "BUILD_COMPLETE";
        case msg.IShipyard_Status.BUILD_CANCELED: return "BUILD_CANCELED";
        case msg.IShipyard_Status.BUILD_FROZEN: return "BUILD_FROZEN";
        case msg.IShipyard_Status.BUILD_FAILED: return "BUILD_FAILED";
        case msg.IShipyard_Status.BLUEPRINT_NOT_FOUND: return "BLUEPRINT_NOT_FOUND";
        default: return undefined;
    }
}
